.pre-commit-config.yaml

```diff
@@ -1,4 +1,11 @@
 repos:
+  - repo: https://github.com/astral-sh/ruff-pre-commit
+    rev: v0.0.287
+    hooks:
+      - id: ruff
+        args: [--exit-non-zero-on-fix]
```

Comment:
How about adding ``--diff`` and ``--format github``?

The former will write the diff of the proposed fix to the logs, and the latter should output check annotations for GitHub for PRs

```suggestion
        args: ['--exit-non-zero-on-fix', '--diff', '--format', 'github']
```

Reply:
Ref: --diff --format github

Reply:
Hmm, I added them (re-ordered, without quotes, and combined `--format, github` into a single argument) but Ruff didn't fail as expected?

Expected:
* https://github.com/python/cpython/actions/runs/6130878966/job/16640392499?pr=109161#step:4:126

Actual:
* https://github.com/python/cpython/actions/runs/6136692710/job/16651782392?pr=109161#step:4:126
* https://github.com/hugovk/cpython/actions/runs/6136706649/job/16651813966?pr=109161#step:4:126

Reply:
Caused by `--diff`, I've removed it.

Reply:
With `--format=github`:

```console
❯ pre-commit run --all-files ruff
ruff.....................................................................Failed
- hook id: ruff
- exit code: 1

::error title=Ruff (F811),file=/Users/hugo/github/cpython/Lib/test/test_binascii.py,line=279,col=9,endLine=279,endColumn=27::Lib/test/test_binascii.py:279:9: F811 Redefinition of unused `test_hex_roundtrip` from line 236
::error title=Ruff (F811),file=/Users/hugo/github/cpython/Lib/test/test_monitoring.py,line=1211,col=9,endLine=1211,endColumn=35::Lib/test/test_monitoring.py:1211:9: F811 Redefinition of unused `test_line_then_instruction` from line 1188
::error title=Ruff (F811),file=/Users/hugo/github/cpython/Lib/test/test_monitoring.py,line=1216,col=9,endLine=1216,endColumn=35::Lib/test/test_monitoring.py:1216:9: F811 Redefinition of unused `test_instruction_then_line` from line 1193
```

The annotations are good for PRs:

<details>
<summary>GitHub screenshot</summary>

<img width="1110" alt="image" src="https://github.com/python/cpython/assets/1324225/4dcb25a7-f0e7-45ef-904b-36963bd63bfc">

</details>

Without `--format=github` (not shown here, filename is bold, F811 is red):

```console
❯ pre-commit run --all-files ruff
ruff.....................................................................Failed
- hook id: ruff
- exit code: 1

Lib/test/test_binascii.py:279:9: F811 Redefinition of unused `test_hex_roundtrip` from line 236
Lib/test/test_monitoring.py:1211:9: F811 Redefinition of unused `test_line_then_instruction` from line 1188
Lib/test/test_monitoring.py:1216:9: F811 Redefinition of unused `test_instruction_then_line` from line 1193
Found 3 errors.
```

<details>
<summary>Console screenshot</summary>

<img width="762" alt="image" src="https://github.com/python/cpython/assets/1324225/cab03d85-c347-4b1a-bfd5-b77152feee16">


</details>


It would be nice to have the GitHub formatting just for PRs and not for local use. 

Is it possible to apply `--format=github` just via the workflow? Like with an env var or something?

Reply:
Yes, `RUFF_FORMAT=github` should do it:
```
--format <FORMAT>
    Output serialization format for violations [env: RUFF_FORMAT=] [possible values: text, json, json-lines, junit, grouped, github, gitlab, pylint, azure]
```

Reply:
The annotations on this PR look great -- this was a fab idea @AA-Turner!